feature: 2 algorithms Website-specific workers and Round robin +DNS

deputinizer opened the pull request:
They use a lot less resources, since eg.
1 MB buffer is assigned to the same URL
and other 1 kB buffer is assigned to another, constant URL

Also spamming goroutines with httpClient.DoContext was unwise.
Runtime goroutine switching took a lot of CPU.

Currently, i see a lot less RAM+CPU usage after removing task-issuing channel.

deputinizer committed:
feat: revert back to Website-specific workers
The use a lot less resources, since eg.
1 MB buffer is assigned to the same URL
and other 1 kB buffer is assigned to another, constant URL

Also spamming goroutines with httpClient.DoContext was unwise.
Runtime goroutine switching took a lot of CPU.

deputinizer commented:
Also, added stats `localhost:8049/downloaded`
```csv
 4244.084 MB	https://www.moex.com
 2412.250 MB	https://mail.ru
 2330.066 MB	https://www.wildberries.ru
 2307.288 MB	https://abcobmen.com
 2106.000 MB	https://www.rambler.ru
 1780.000 MB	https://www.yaplakal.com
  595.322 MB	https://www.avito.ru
  542.826 MB	https://any.yandex.ru
  529.330 MB	https://montreal.mid.ru
```

deputinizer commented:
<details>
  <summary>Main CPU abuser now [spoiler]</summary>
  
![image](https://user-images.githubusercontent.com/100740281/156845947-08c5fad0-c9ae-4573-8b02-92de9c056241.png)
  
</details>

deputinizer commented:
<details>
  <summary>We should nuke this buffer with Bayraktars</summary>

![image](https://user-images.githubusercontent.com/100740281/156846592-d1224bbe-004e-4e02-b527-493b7fb48a2f.png)
  
</details>

erkexzcx commented:
I don't like the idea to bring per-website workers. There was an issue that even 1 worker (per website) used all traffic, so workers pool was the way to fix it.

I understand that this way we could save more resources and make this app more performant, but what can we offer for users where even 1 worker per website is too much?

Shall we ignore such users and go ahead with workers-per-website approach?

Or shall we still use the same "workers pool" idea and divide workers equally to each website (rouding workers number so each website has the same amount of workers)?

Opinions? :)

deputinizer commented:
> Shall we ignore such users and go ahead with workers-per-website approach?

Yes

bytes/s > packets/s

Or maybe add 2 options - your worker issuing channel algorithm and worker-per-website algorithm.

deputinizer commented:
Now fasthttp is able to `io.Discard` :smiley: https://github.com/deputinizer/fasthttp/commit/80e5eb0ab1f4788e5b0e375dd3f4d03d6879ecab

deputinizer committed:
chore: add fasthttp+discard, go get -u all

deputinizer commented:
<details>
  <summary>Discards like a boss (heap pprof)</summary>
  
![image](https://user-images.githubusercontent.com/100740281/156852287-c042221e-540b-4ad5-9850-225e9f3926c0.png)

  
</details>

deputinizer commented:
<details>
  <summary>There's something wrong i think (CPU pprof)</summary>
  
![image](https://user-images.githubusercontent.com/100740281/156852467-3fbf44fb-3ce4-4d15-bec1-0bf7979e29a2.png)

</details>

deputinizer committed:
chore: bump fasthttp v1.33.0-discard2

deputinizer committed:
feat: discard http body

deputinizer committed:
fix: website counters with discard

deputinizer committed:
feat: allow both algorithms
- Round robin with websites issued from channel
- Parallel goroutines with constant website

deputinizer commented:
So, looks like:
- round robin prioritizes slower websites and spam with `net.Dial`
- parallel goroutines (few per website) prioritize wasting throughput

You can choose which task issuing algorithm to use with a flag:
`-roundrobin=true`

deputinizer committed:
chore: bump fasthttp-discard3

deputinizer committed:
chore: bump fashttp+discard4

deputinizer commented:
Ok, so stats now work with `-roundrobin=true`
```
   43.239 MB	https://abcobmen.com
   33.885 MB	https://www.rambler.ru
   22.483 MB	https://mail.ru
   19.784 MB	https://pikabu.ru
   16.967 MB	https://www.moex.com
    6.215 MB	https://montreal.mid.ru
    2.487 MB	https://www.avito.ru
    1.695 MB	https://www.yaplakal.com
```

For comparison `-roundrobin=false`
```
  207.586 MB	https://www.moex.com
  128.076 MB	https://www.yaplakal.com
  119.014 MB	https://www.wildberries.ru
  118.598 MB	https://mail.ru
  113.556 MB	https://www.rambler.ru
  109.937 MB	https://abcobmen.com
   50.105 MB	https://www.avito.ru
   32.551 MB	https://any.yandex.ru
   26.804 MB	https://pikabu.ru
   25.712 MB	https://montreal.mid.ru
   19.033 MB	https://ww-pay.net
   14.681 MB	https://superchange.net
```

deputinizer commented:
I suggest `-roundrobin=false` by default
since new StopPropaganda's users are going to be astonished after seeing 10 Gbit/s on their machines on first try. :satisfied:

deputinizer committed:
fix: high intensity errors

deputinizer committed:
feat: add dependency golang.org/x/net/dns/dnsmessage

deputinizer committed:
feat: copy basic dns Resolver from golang internals

deputinizer committed:
feat: more dns from internals

deputinizer committed:
refactor: unix-like because we don't want /etc/resolv.conf
... but a reasonable implementation of it

deputinizer committed:
feat: make dns config public

deputinizer committed:
feat: mark all errors in DNS client with [SP]

deputinizer committed:
feat: more code from golang internals

deputinizer committed:
feat: trying to use custom implementation of dns...

deputinizer committed:
feat: add MasterStopPropagandaResolver with spdnsclient
unix-like DNS client but without /etc/resolv.conf
Just pure hardcoded config and []string

deputinizer committed:
fix: delete injectionresolver hack
Android started working because of this hack (probably).
But now, we have a whole /etc/resolv.conf alternative,
so it should work even better.

deputinizer commented:
Ok, this DNS client replaces `/etc/resolv.conf` in a proper way (no `net.Resolver.Dial` hacks), just a lot of copy paste.
I had to `ctrl+C` `ctrl+V` a lot of unix DNS client, because 90% of the code was private :cry:

deputinizer commented on internal/stoppropaganda/customresolver/customresolver.go:
```diff
@@ -6,11 +6,34 @@ import (
+		Timeout:  5 * time.Second,
+		Attempts: 2,
+	}
+	conf.Servers = targets.ReferenceDNSServersForHTTP
```
Here is the actual fake `/etc/resolv.conf`

deputinizer committed:
fix: building on all OSs

erkexzcx committed:
rm: redundant return

erkexzcx committed:
refactor: move packages from stoppropaganda to internal

erkexzcx committed:
rm: unused code

erkexzcx committed:
feat: use algorithm name instead

erkexzcx committed:
fix: consistent description

erkexzcx committed:
refactor: README.md documentation, set default to 'fair' alg

erkexzcx committed:
refactor: web ui documentation

erkexzcx commented:
Regarding https://github.com/erkexzcx/stoppropaganda/pull/103/commits/30f0113715b9b1cc9b8f9ada345cd1588a64ac2a - this was not needed. All the blockings were done by `if !website.allowedToRun()` part.

erkexzcx committed:
rm: blocking code

erkexzcx commented:
Works great on Windows 10 :+1:
